File: apps/web/src/app/schemas/plantAssembly.ts
import { z } from "zod";

export const coords2dSchema = z.object({
  x: z.number(),
  y: z.number(),
});

export const assetInstanceSchema = z.object({
  asset_id: z.string(),
  component_type_id: z.string(),
  display_name: z.string(),
  position_2d: coords2dSchema,
  position_3d: z.object({ x: z.number(), y: z.number(), z: z.number() }).optional(),
  configured_ports: z.array(z.string()).default([]),
  configured_signals: z.array(z.string()).default([]),
  overrides: z.record(z.unknown()).default({}),
  enabled_fault_modes: z.array(z.string()).default([]),
});

export const plantConnectionSchema = z.object({
  connection_id: z.string(),
  from_asset_id: z.string(),
  from_port_id: z.string(),
  to_asset_id: z.string(),
  to_port_id: z.string(),
  connection_kind: z.enum(["power", "signal", "mechanical", "airflow", "fluid", "mounting", "data"]),
  approved: z.boolean(),
  lag_min_ms: z.number().int().nonnegative(),
  lag_max_ms: z.number().int().nonnegative(),
  notes: z.string().optional().default(""),
});

export const plantAssemblySchema = z.object({
  assembly_id: z.string(),
  plant_id: z.string(),
  version: z.string(),
  assets: z.array(assetInstanceSchema),
  connections: z.array(plantConnectionSchema),
  global_tags: z.array(z.string()).default([]),
  metadata: z.record(z.unknown()).default({}),
});

export type Coords2D = z.infer<typeof coords2dSchema>;
export type AssetInstance = z.infer<typeof assetInstanceSchema>;
export type PlantConnection = z.infer<typeof plantConnectionSchema>;
export type PlantAssembly = z.infer<typeof plantAssemblySchema>;

export interface CompatibilityResult {
  compatible: boolean;
  severity: "ok" | "warning" | "error";
  reason: string;
  warnings: string[];
  required_adapters: string[];
  from_medium?: string;
  to_medium?: string;
  from_quantity_kind?: string;
  to_quantity_kind?: string;
}

File: apps/api/app/schemas/plant_assembly.py
"""Pydantic mirrors for plant assembly contracts."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class Coords2D(BaseModel):
    model_config = ConfigDict(extra="forbid")
    x: float
    y: float


class Coords3D(BaseModel):
    model_config = ConfigDict(extra="forbid")
    x: float
    y: float
    z: float


class AssetInstance(BaseModel):
    model_config = ConfigDict(extra="forbid")

    asset_id: str
    component_type_id: str
    display_name: str
    position_2d: Coords2D
    position_3d: Coords3D | None = None
    configured_ports: list[str] = Field(default_factory=list)
    configured_signals: list[str] = Field(default_factory=list)
    overrides: dict[str, Any] = Field(default_factory=dict)
    enabled_fault_modes: list[str] = Field(default_factory=list)


ConnectionKind = Literal["power", "signal", "mechanical", "airflow", "fluid", "mounting", "data"]


class PlantConnection(BaseModel):
    model_config = ConfigDict(extra="forbid")

    connection_id: str
    from_asset_id: str
    from_port_id: str
    to_asset_id: str
    to_port_id: str
    connection_kind: ConnectionKind
    approved: bool
    lag_min_ms: int = 0
    lag_max_ms: int = 0
    notes: str = ""


class PlantAssembly(BaseModel):
    model_config = ConfigDict(extra="forbid")

    assembly_id: str
    plant_id: str
    version: str
    assets: list[AssetInstance] = Field(default_factory=list)
    connections: list[PlantConnection] = Field(default_factory=list)
    global_tags: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class CheckConnectionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    from_component_type_id: str
    from_port_id: str
    to_component_type_id: str
    to_port_id: str


class ValidateAssemblyRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    plant_assembly: PlantAssembly
    component_library: dict[str, Any] | None = None


class AnalyzeAssemblyRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    plant_assembly: PlantAssembly
    component_library: dict[str, Any] | None = None


class ScoreFaultsRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    plant_assembly: PlantAssembly
    observed_signals: dict[str, Any] = Field(default_factory=dict)
    data_quality: dict[str, Any] = Field(default_factory=dict)
    component_library: dict[str, Any] | None = None